"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FailsafeTimer_exports = {};
__export(FailsafeTimer_exports, {
  FailsafeTimer: () => FailsafeTimer,
  MatterFabricConflictError: () => MatterFabricConflictError
});
module.exports = __toCommonJS(FailsafeTimer_exports);
var import_general = require("#general");
/**
 * @license
 * Copyright 2022-2025 Matter.js Authors
 * SPDX-License-Identifier: Apache-2.0
 */
class MatterFabricConflictError extends import_general.MatterFlowError {
}
const logger = import_general.Logger.get("FailsafeTimer");
class FailsafeTimer {
  constructor(associatedFabric, expiryLengthSeconds, maxCumulativeFailsafeSeconds, expiryCallback) {
    this.associatedFabric = associatedFabric;
    this.#expiryCallback = expiryCallback;
    this.#failsafeTimer = this.#startFailsafeTimer(expiryLengthSeconds);
    this.#maxCumulativeFailsafeTimer = import_general.Time.getTimer(
      "Max cumulative failsafe",
      maxCumulativeFailsafeSeconds * 1e3,
      () => this.expire()
    ).start();
  }
  #expiryCallback;
  #failsafeTimer;
  #maxCumulativeFailsafeTimer;
  async close() {
    if (this.#failsafeTimer.isRunning) {
      this.#failsafeTimer.stop();
    }
    if (this.#maxCumulativeFailsafeTimer.isRunning) {
      this.#maxCumulativeFailsafeTimer.stop();
    }
  }
  /** Handle "Re-Arming" an existing FailSafe context to extend the timer, expire or fail if not allowed. */
  async reArm(associatedFabric, expiryLengthSeconds) {
    if (!this.#failsafeTimer.isRunning) {
      throw new import_general.MatterFlowError("FailSafe already expired.");
    }
    if (this.associatedFabric?.fabricIndex !== associatedFabric?.fabricIndex) {
      throw new import_general.MatterFlowError(
        `FailSafe already armed (index=${this.associatedFabric?.fabricIndex}) with different fabric (index=${associatedFabric?.fabricIndex}).`
      );
    }
    this.#failsafeTimer.stop();
    if (expiryLengthSeconds === 0) {
      await this.expire();
    } else {
      this.#failsafeTimer = this.#startFailsafeTimer(expiryLengthSeconds);
    }
  }
  /** Expire the FailSafe context. This is called by the timer and can also be called manually if needed. */
  async expire() {
    this.complete();
    await this.#expiryCallback();
  }
  /** Complete the FailSafe context. This is called when the commissioning is completed. */
  complete() {
    this.#failsafeTimer.stop();
    this.#maxCumulativeFailsafeTimer.stop();
  }
  #startFailsafeTimer(expiryLengthSeconds) {
    return import_general.Time.getTimer(
      "Failsafe expiration",
      expiryLengthSeconds * 1e3,
      () => this.expire().catch((e) => logger.error("Error during failsafe expiration", e))
    ).start();
  }
}
//# sourceMappingURL=FailsafeTimer.js.map
